linkerd/main/src/main/scala/io/buoyant/linkerd/Main.scala

```diff
@@ -28,28 +27,47 @@ object Main extends App {
 
     args match {
       case Array(path) =>
-        val config = loadLinker(path)
-        val linker = config.mk()
-        val admin = initAdmin(config, linker)
-        val telemeters = linker.telemeters.map(_.run())
-        val routers = linker.routers.map(initRouter(_))
-
-        log.info("initialized")
-        registerTerminationSignalHandler(config.admin.flatMap(_.shutdownGraceMs))
-        closeOnExit(Closable.sequence(
-          Closable.all(routers: _*),
-          Closable.all(telemeters: _*),
-          Closable.all(admin: _*)
-        ))
-        Await.all(routers: _*)
-        Await.all(telemeters: _*)
-        Await.all(admin: _*)
-
+        start(loadLinker(path))(TerminationHook) match {
```

Comment:
I find this refactor to be much harder to follow and understand than what was here before.  I'm guessing that you did this refactor to facilitate testing?  Unfortunately, it looks like the test doesn't really test the signal handling at all so I'm not sure it's worth refactoring the code for such a test.

Ideally we would be able to leave the structure of the code the same and test the behavior with `Signal.raise` but I'm not sure if this is possible within a unit test or how the testing framework handles it.  If such a test is not feasible, I think it's okay to resort to manual testing for features like this.

Reply:
Yes, I tried hard to come up with something the can be testable but it felt the same to me.
It makes the code unreadable.
I am happy to close the pull request until we find a better strategy  for it.

Reply:
I think that would be my preference.  This kind of thing is, unfortunately, pretty tricky to test.  I do appreciate your work on this anyway.